src/astarte-charts/react/hooks/useChartProviders.ts

```diff
@@ -0,0 +1,103 @@
+/*
+   This file is part of Astarte.
+
+   Copyright 2020 Ispirata Srl
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import { useCallback, useEffect, useState } from 'react';
+import type {
+  AstarteChartProvider,
+  AstarteChartSingleValue,
+  AstarteChartListValue,
+} from 'astarte-charts';
+
+type Status = 'loading' | 'ok' | 'error';
+
+type DataFetcher<Data> =
+  | {
+      status: 'loading';
+      data: Data | null;
+      error: Error | null;
+      refresh: () => Promise<void>;
+    }
+  | {
+      status: 'ok';
+      data: Data;
+      error: Error | null;
```

Comment:
That's a great point, thanks for raising it.
The approach where we have data and an error (lastError) at the same time confuses the user and is trickier for the chart to act upon the error.
The approach where we have data XOR error is also bad since a common case is that the chart wants to always show the latest data, optionally notifying an error.
So I reworked this hook with a third approach:
- it will always provide the latest data to the chart, if available
- it reports whether new data is being fetched
- the chart can pass a `onError` callback to handle errors how it sees fit